tests/unit/lib/SGMatrix_unittest.cc

```diff
@@ -761,3 +761,37 @@ TEST(SGMatrixTest,iterator)
 	for (auto v: mat)
 		EXPECT_EQ(mat[index++], v);
 }
+
+TEST(SGMatrixTest,column_iterator)
+{
+	constexpr index_t size=5;
+	SGMatrix<float64_t> mat(size, size);
+	linalg::range_fill(mat, 1.0);
+
+	SGMatrix<float64_t>::column_iterator col_itr(mat);
+
+	EXPECT_EQ(mat.get_column(0),*col_itr);
+	++col_itr;
+	EXPECT_EQ(mat.get_column(1),*col_itr);
+	--col_itr;
+	EXPECT_EQ(mat.get_column(0),*col_itr++);
+	++col_itr;
+	EXPECT_EQ(mat.get_column(2),*col_itr--);
+	--col_itr;
+	EXPECT_EQ(mat.get_column(0),*col_itr);
+	col_itr += 2;
+	EXPECT_EQ(mat.get_column(2),*col_itr);
+	col_itr -= 1;
+	EXPECT_EQ(mat.get_column(1),*col_itr);
+	EXPECT_EQ(mat.get_column(2),col_itr[1]);
+	
+	
+	auto new_itr = col_itr + 2;
+	EXPECT_EQ(mat.get_column(3),*new_itr);
+
+	//range-based loop should work as well
+	int idx;
+	std::for_each(mat.begin_column(),mat.end_column(),[&idx,mat](auto* vec){ EXPECT_EQ(*vec,mat.get_column(idx++));});
```

Comment:
This doesn't work and throws this: https://pastebin.com/qArHsKcU

Reply:
should be `std::for_each(mat.begin_column(),mat.end_column(),[&idx,mat](const auto& vec){ EXPECT_EQ(*vec,mat.get_column(idx++));}); `
You are passing a non-pointer type to a lambda expecting a pointer

Reply:
Okay using that gives https://pastebin.com/XVvURTpb. 
I think for_each deferences it for us, so we don't need to use ```*v``` again. It compiles succesfully when we use ```v```

Reply:
You are dereferencing a non-pointer type without an operator* overload. Yes, you have to write it as `v`

Reply:
Right!.